### PR TITLE
Reducing calls to checkConflictingNodes()

### PR DESCRIPTION
**Pull request by heschlie**
## Description
This fix should result in less calls to `checkConflictingNodes()` which makes API calls to `client.Nodes.List()` which is a very expensive call.  This should result in less strain during large clusters with node churn.

I'm not sure if this should have a release note tied to it.  

## Release Notes
```release-note
calico/node will now only check for conflicting Node IPs when initially getting an IP or when a change in IP is detected.  This should reduce the load on the cluster when a large number of nodes are restarting.
```

### CODE REVIEW

File: calico_node/startup/startup.go [hunk 1]
```diff
@@ -295,8 +296,12 @@ func configureIPsAndSubnets(node *api.Node) {
 		adm := os.Getenv("IP6_AUTODETECTION_METHOD")
 		cidr := autoDetectCIDR(adm, 6)
 		if cidr != nil {
-			// We autodetected an IPv6 address so update the value in the node.
-			node.Spec.BGP.IPv6Address = cidr
+			// We autodetected an IPv6 address, if we have no previous IP, or it is different than the previous IP,
+			// we need to update it and check for conflicts.
+			if node.Spec.BGP.IPv6Address == nil || !node.Spec.BGP.IPv6Address.IP.Equal(cidr.IP) {
+				node.Spec.BGP.IPv6Address = cidr
+				checkConflictingNodes(client, node)
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: you potentially call this twice - once for ipv4 and once for ipv6.

Comment (heschlie) [inline calico_node/startup/startup.go]: Gah you're right, which kinda defeats the purpose, will refactor it to only call once

File: calico_node/startup/startup.go [hunk 2]
```diff
@@ -84,14 +84,11 @@ func main() {
 	node := getNode(client, nodeName)
 
 	// Configure and verify the node IP addresses and subnets.
-	configureIPsAndSubnets(node)
+	configureIPsAndSubnets(client, node)
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: minor nit - rather than pass the client here it might be nicer to make it so `configureIPsAndSubnets` returns whether or not the an IP on the node changed, and we can use it below.

File: calico_node/startup/startup.go [hunk 3]
```diff
@@ -267,8 +268,12 @@ func configureIPsAndSubnets(node *api.Node) {
 		adm := os.Getenv("IP_AUTODETECTION_METHOD")
 		cidr := autoDetectCIDR(adm, 4)
 		if cidr != nil {
-			// We autodetected an IPv4 address so update the value in the node.
-			node.Spec.BGP.IPv4Address = cidr
+			// We autodetected an IPv4 address, if we have no previous IP, or it is different than the previous IP,
+			// we need to update it and check for conflicts.
+			if node.Spec.BGP.IPv4Address == nil || !node.Spec.BGP.IPv4Address.IP.Equal(cidr.IP) {
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: Hm, this only checks the IP part of the CIDR, not the netmask.

I think we might want to set `node.Spec.BGP.IPv4Address = cidr` always - it's possible that the IP stays the same but the subnet changes.

**Comment by caseydavenport** [comment]
In the original issue, there was also a suggestion that we add a config option to disable the check entirely - @heschlie could you add that to this PR?

**Comment by caseydavenport** [comment]
@heschlie we'll need a corresponding docs PR to document the new configuration option.  Could you link to it here once you've created it?

**Comment by heschlie** [comment]
Does it need a separate PR or could/should it just be in this PR?

**Comment by caseydavenport** [comment]
@heschlie oops, forgot we're in the Calico repo now!

It can share this PR.

File: calico_node/startup/startup.go [hunk 4]
```diff
@@ -267,7 +274,11 @@ func configureIPsAndSubnets(node *api.Node) {
 		adm := os.Getenv("IP_AUTODETECTION_METHOD")
 		cidr := autoDetectCIDR(adm, 4)
 		if cidr != nil {
-			// We autodetected an IPv4 address so update the value in the node.
+			// We autodetected an IPv4 address, if we have no previous IP, or it is different than the previous IP,
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: Sorry I didn't notice before, but this check is not quite sufficient as it only covers the auto-detection case.

I think we also need to handle the case when the IP address has been explicitly specified in the environment.

Comment (heschlie) [inline calico_node/startup/startup.go]: Yea that's true, I'll add similar checks there as well.

Comment (robbrockbank) [inline calico_node/startup/startup.go]: perhaps easier to store the initial IPv4 and IPv6 addresses and do the comparison once at the end of the fn?

File: calico_node/startup/startup.go [hunk 5]
```diff
@@ -314,11 +333,16 @@ func configureIPsAndSubnets(node *api.Node) {
...
 		if ipv6Env != "" {
-			node.Spec.BGP.IPv6Address = parseIPEnvironment("IP6", ipv6Env, 6)
+			envIp6 := parseIPEnvironment("IP6", ipv6Env, 6)
+			if node.Spec.BGP.IPv6Address == nil || !node.Spec.BGP.IPv6Address.IP.Equal(envIp6.IP) {
+				checkConflicts = true
+			}
+			node.Spec.BGP.IPv6Address = envIp6
 		}
 		validateIP(node.Spec.BGP.IPv6Address)
 	}
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: Would it be nicer and less code to store off the original value at the top of the function, and compare it to the new value at the bottom?

e.g.

```
oldv4 := node.Spec.BGP.IPv4Address
oldv6 := ...

// Rest of function.
...

if node.Spec.BGP.IPv4Addres != oldv4 ||  node.Spec.BGP.IPv6Address != oldv6 {
  ...
}
```

Comment (robbrockbank) [inline calico_node/startup/startup.go]: Yup I think I suggested the same thing earlier :-)

Comment (caseydavenport) [inline calico_node/startup/startup.go]: Ah, missed that :)

Just driving by.

**Comment by heschlie** [comment]
@caseydavenport @robbrockbank I think this should be good to go now, docs for new env var are in and the `configureIPsAndSubnets()` call simplified.

File: calico_node/startup/startup.go (hunk 4 above)
Comment (caseydavenport) [inline calico_node/startup/startup.go]: I think this comment is wrong now that we've moved the check outside of this branch - should probably just be reverted to what it was.

File: calico_node/startup/startup.go [hunk 6]
```diff
@@ -274,7 +285,8 @@ func configureIPsAndSubnets(node *api.Node) {
 		}
 	} else {
 		if ipv4Env != "" {
-			node.Spec.BGP.IPv4Address = parseIPEnvironment("IP", ipv4Env, 4)
+			envIp := parseIPEnvironment("IP", ipv4Env, 4)
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: I think we should revert this change - not required.

File: calico_node/startup/startup.go (hunk 4 above)
Comment (caseydavenport) [inline calico_node/startup/startup.go]: Same for a few other comments in this PR.

File: calico_node/startup/startup.go (hunk 6 above)
Comment (caseydavenport) [inline calico_node/startup/startup.go]: Same for IPv6 below.

**Comment by caseydavenport** [comment]
@heschlie I think we need a unit test for this.

File: calico_node/startup/startup.go [hunk 7]
```diff
@@ -303,11 +316,21 @@ func configureIPsAndSubnets(node *api.Node) {
...
+			node.Spec.BGP.IPv6Address = envIp6
 		}
 		validateIP(node.Spec.BGP.IPv6Address)
 	}
 
+	// Detect if we've seen the IP address change, and flag that we need to check for conflicting Nodes
+	if oldIpv4 == nil || !node.Spec.BGP.IPv4Address.IP.Equal(oldIpv4.IP) {
+		checkConflicts = true
+	}
+	if oldIpv6 == nil || !node.Spec.BGP.IPv6Address.IP.Equal(oldIpv6.IP) {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: For an IPv4 deployment, oldIpv6 will always be nil every time we start, so we'll always perform the conflict check.

This should be:
```
if (oldIpv6 == nil && node.Spec.BGP.IPv6Address != nil) || (oldIpv6 != nil && !node.Spec.BGP.IPv6Address.IP.Equal(oldIpv6.IP))
```

or something like that

File: calico_node/startup/startup.go [hunk 8]
```diff
@@ -303,11 +316,21 @@ func configureIPsAndSubnets(node *api.Node) {
...
-			node.Spec.BGP.IPv6Address = parseIPEnvironment("IP6", ipv6Env, 6)
+			envIp6 := parseIPEnvironment("IP6", ipv6Env, 6)
+			node.Spec.BGP.IPv6Address = envIp6
 		}
 		validateIP(node.Spec.BGP.IPv6Address)
 	}
 
+	// Detect if we've seen the IP address change, and flag that we need to check for conflicting Nodes
+	if oldIpv4 == nil || !node.Spec.BGP.IPv4Address.IP.Equal(oldIpv4.IP) {
+		checkConflicts = true
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: nit:  Do you really need to define checkConflicts?  You could just return true here.

File: calico_node/startup/startup.go [hunk 9]
```diff
@@ -303,11 +316,21 @@ func configureIPsAndSubnets(node *api.Node) {
...
 		if ipv6Env != "" {
-			node.Spec.BGP.IPv6Address = parseIPEnvironment("IP6", ipv6Env, 6)
+			envIp6 := parseIPEnvironment("IP6", ipv6Env, 6)
+			node.Spec.BGP.IPv6Address = envIp6
 		}
 		validateIP(node.Spec.BGP.IPv6Address)
 	}
 
+	// Detect if we've seen the IP address change, and flag that we need to check for conflicting Nodes
+	if oldIpv4 == nil || !node.Spec.BGP.IPv4Address.IP.Equal(oldIpv4.IP) {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: You should log in these two branches.

File: master/reference/node/configuration.md [hunk 10]
```diff
@@ -16,6 +16,7 @@ The `calico/node` container is primarily configured through environment variable
...
 | IP_AUTODETECTION_METHOD| The method to use to autodetect the IPv4 address for this host. This is only used when the IPv4 address is being autodetected. See [IP Autodetection methods](#ip-autodetection-methods) for details of the valid methods. | string | first-found |
 | IP6_AUTODETECTION_METHOD| The method to use to autodetect the IPv6 address for this host. This is only used when the IPv6 address is being autodetected. See [IP Autodetection methods](#ip-autodetection-methods) for details of the valid methods. | string | first-found |
+| DISABLE_NODE_IP_CHECK| Skips checks for duplicate Node IPs. This can reduce the load on the cluster when a large number of Nodes are restarting. | string | false |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: string -> boolean

File: calico_node/startup/startup.go [hunk 11]
```diff
@@ -84,14 +84,17 @@ func main() {
 	node := getNode(client, nodeName)
 
 	// Configure and verify the node IP addresses and subnets.
-	configureIPsAndSubnets(node)
+	checkConflicts := configureIPsAndSubnets(node)
+
+	// If we report an IP change (v4 or v6) we should verify there are no
+	// conflicts between Nodes.
+	if checkConflicts && os.Getenv("DISABLE_NODE_IP_CHECK") != "true" {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: It feels like we should be more forgiving about the value that people can pass in - I think anything that resembles a boolean true should be sufficient:  true t y 1 (case insensitive)

@caseydavenport @heschlie WDYT?

Comment (caseydavenport) [inline calico_node/startup/startup.go]: What do we respect for other true/false variables?  Seems fine/good to do, but I'm happy to merge without.

File: calico_node/tests/st/bgp/test_backends.py [hunk 12]
```diff
@@ -56,7 +56,7 @@ def test_bgp_backends(self):
             workload_host3 = host3.create_workload("workload3", network=network1, ip=DEFAULT_IPV4_ADDR_3)
 
             # Allow network to converge
-            self.assert_true(workload_host1.check_can_ping(workload_host2.ip, retries=10))
+            self.assert_true(workload_host1.check_can_ping(workload_host2.ip, retries=15))
```
Comment (caseydavenport) [inline calico_node/tests/st/bgp/test_backends.py]: hmm.. I wonder why these changes are required.. have we regressed some sort of timing thing? Any ideas @robbrockbank?

File: calico_node/startup/startup.go (hunk 11 above)
Comment (robbrockbank) [inline calico_node/startup/startup.go]: Yes I'm happy to merge without too.  I'll raise an issue to track better handling of boolean envs.

File: calico_node/tests/st/bgp/test_backends.py (hunk 12 above)
Comment (robbrockbank) [inline calico_node/tests/st/bgp/test_backends.py]: No idea - these should not need to be updated for this PR.  Perhaps local environment is slower?

Comment (heschlie) [inline calico_node/tests/st/bgp/test_backends.py]: I was trying to reduce the number of inconsistent test failures I was seeing, will revert as I don't think it made any difference.

**Comment by heschlie** [comment]
I've Added in some UT's in a table driven test in the `startup_test.go`

File: calico_node/startup/startup.go [hunk 13]
```diff
@@ -236,7 +239,7 @@ func getNode(client *client.Client, nodeName string) *api.Node {
 
 // configureIPsAndSubnets updates the supplied node resource with IP and Subnet
 // information to use for BGP.
-func configureIPsAndSubnets(node *api.Node) {
+func configureIPsAndSubnets(node *api.Node) bool {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: One final thing:  I think the comment should indicate that it returns true if the node IP addresses are updated.

**Comment by heschlie** [comment]
I've updated the UT to do some IPv6 checks as well.

**Comment by caseydavenport** [comment]
Oh, and could you tweak the release note to talk about "calico/node" instead of the startup script?  That's more user-facing.
